### PR TITLE
Adds sentient spiders to midround events

### PR DESCRIPTION
**Pull request by benbot16**
## About The Pull Request

This is a partial port of tgstation/tgstation#54451.

This PR replaces the AI-controlled spider event with an event that spawns sentient spiders. It also reworks how spiders function a bit.

The event itself works by spawning 2 midwife spiders, using the same spawn mechanics as xenomorphs. These midwives immediately get the ability to lay a cluster of eggs, giving them some initial numbers, since midwives are squishy.

Egg clusters themselves have also been changed. They now don't hatch into spiderlings, instead allowing ghosts to interact with them after a minute to become a spider of their choice. If a spider eats a human, the spiders available will be one more powerful spider (tarantula/midwife/viper) and 3 less powerful spiders (base spider/nurse/hunter). If they eat something that's not a human, it'll just be 4 less-powerful spiders.

There's also been a few changes to spiders themselves. Spiders have been sped up, but they now slow down as they take damage. Nurses and midwives have also been given the ability to heal other spiders. AI-controlled spiders are also generally slower than player-controlled ones. Spiders also can't speak common anymore, instead speaking buzzwords, but they can still understand common. 

Additionally, spiders now have a new type of venom. This venom does less damage, but will deal stamina damage over time based on how much is in someone's system. It also has a chance to paralyze, and paralyzes more rapidly if there's purgatives in someone's system. 

New values for each spider type:
Giant spider: 100 health, 25 damage, 1 speed, -0.1 on web, 0 poison
Hunter: 80 health, 20 damage, 0 speed, 5 poison
Viper: 50 health, 1 damage, 0 speed, 12 poison (venom)
Nurse: 40 health, 10 damage, 1 speed, 0.5 on web, 3 poison, faster webbing
Midwife: 80 health, 10 damage, 2 speed, 1 on web, 3 poison, fastest webbing
Tarantula: 300 health, 35 damage, 5 speed off web, 0 speed on web, fast webbing

## Why It's Good For The Game

The spider infestation event is really lame right now, and this would improve it. This also improves spider AI a bit, so they'll be a bit more interesting to deal with when exploring or dealing with spiderlings created by other things.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>
The event:

https://user-images.githubusercontent.com/9423435/194977423-546a1b6c-a778-4f48-862b-f9d76021c795.mp4

Eggs:

https://user-images.githubusercontent.com/9423435/194977468-344adcd1-e664-453f-9f6c-39352756f035.mp4

Spider healing:

https://user-images.githubusercontent.com/9423435/194977695-9e5f6ad7-c803-49cc-8339-c315275b02bd.mp4

</details>

## Changelog
:cl: IndieanaJones, Rukofamicom, Benbot16
add: The Spider infestation midround event now spawns sentient spiders. These fall under the "spider" midround preference.
add: Nurse spiders can now heal other spiders.
tweak: Webs can now be stacked, up to a limit of 3.
tweak: Reworks how spider eggs spawn new spiders.
tweak: Spiders now take damage slowdown below 75% health.
tweak: Spider venom will now cause stamina damage, and paralyzes in large doses or if purged with chemicals.
tweak: Spiders now speak buzzwords. They can still understand common.
tweak: Tweaks spider AI.
balance: Tweaks several spider stats, mostly decreasing health and increasing speed.
/:cl:

### CODE REVIEW

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: this is not consistent with our other enemy lists and doesn't use job name defines

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:725]: make this consistent with xenos midround, there's a log message about execute failure. also use length() instead of len

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Done

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:725]: Done

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: make a ROLE_SPIDER

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: change to 2,2,1,1,1,1,1,0,0,0

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: lower cost to 8

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: do 4

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:707]: add minimum_players = 27

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: len >:( use length()

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Done

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Made one. I'm not entirely sure that I set it up properly, though.

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Done

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:707]: Done

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Done

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Done

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm]:
```suggestion
	maxHealth = 75
	health = 75
```

100 is pretty high as far as HP goes, putting them on par with the average human player. Given just how easily spiders can make more spiders, the average spider should not be on par with a human 1v1 unless you want them to regularly snowball out of control.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm]:
```suggestion
	melee_damage = 15
```

25 is a also a LOT of damage, even when compared to the power of an above average crewmember. 15 is a much better baseline for a "standard" spider, especially if you follow through with making them venomous too.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm:43]:
```suggestion
	poison_per_bite = 4
```

Spiders make a more unique threat if the majority utilize and rely on venom rather than bursty brute force.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm:187]: Everything related to reproduction and feeding should be shifted down and be exclusive to the Broodmother (/midwife)
New Broodmothers should be blocked from spawning via gold slimes and spiderlings, rather than nurses.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm]:
```suggestion
	name = "broodmother"
```

Superior name that better communicates the spider's role, especially to newer players. It's also a name that many players already utilize despite not being the actual name.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm:351]:
```suggestion
	maxHealth = 200
	health = 200
```

Following the logic of making it impossible to produce more broodmothers, they will need a lot more HP given their lacking combat capability.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm]:
```suggestion
	maxHealth = 60
	health = 60
	melee_damage = 15
	poison_per_bite = 6
```

Following the description and changes to the base spider, this follows the intention of a slightly squishier but more venomous variant than the standard giant spider.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm]:
```suggestion
	maxHealth = 35
	health = 35
	melee_damage = 5
	poison_per_bite = 12
```

Lethal in as little as one bite with patience and thought, they deserve to be squishier too. 5 melee_damage means they actually finish a victim paralyzed by venom in a reasonable timeframe while still needing to rely almost entirely on venom in order to kill.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm]:
```suggestion
	maxHealth = 160
	health = 160
	melee_damage = 25
	poison_per_bite = 0
```

As the primary defenders they deserve to have more HP, but 300 is absurdly high for a mob that is intended by design to protect and move in close proximity with others capable of healing it.

**Review by Rukofamicom — CHANGES_REQUESTED** [review]

**Comment by Rukofamicom** [comment]
Even with what I suggested Vipers might still be too extreme on the sliding scale of Standard -> Hunter -> Viper and the curve between them might need to be smoothed out some more. Vipers should play and handle as a high risk, high return choice - not a cheap and easy in-and-out victory. 

I recommend reducing their speed before other stats if they present too much of a problem - glass cannon needs to be reasonably possible to hit in order to be balanced.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (benbot16) [reply code/modules/mob/living/simple_animal/hostile/giant_spider.dm:43]: I intentionally had the "brown" spiders be stronger and not have venom, so that they're differentiated from the hunter/viper spiders. That way it encourages working together a bit more, with the hunters being able to paralyze but not quite having the stopping power of the slower and tankier brown spiders. I'll probably have to tweak damage for both of these more to get to that point, though.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (benbot16) [reply code/modules/mob/living/simple_animal/hostile/giant_spider.dm:187]: I left it on nurses to make make spiders be more decentralized. Having to deal with nurses too removes the possibility of just nuking the couple of broodmothers that exist using a flamer and being done with it.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (benbot16) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm:177]: They already web at a quarter of normal speed.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (benbot16) [reply code/modules/mob/living/simple_animal/hostile/giant_spider.dm]: I left these guys with standard venom, instead of spider venom. They still have one damage because of that. 
I'm not entirely sure how good of an idea it is, but I wanted these guys to still be pretty lethal.

Ideally I'd still like to keep them at this stat level, since their health means that getting hit will put some pretty decent damage slowdown on them, and giving them less health will make it even easier for them to just die to atmos changes without being able to move.

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm
Comment (Rukofamicom) [reply code/modules/mob/living/simple_animal/hostile/giant_spider.dm:177]: I see where that's defined now, my mistake. Can't seem to close these, I may not even be able to - this is my first time leaving any sort of review.

**Comment by Rukofamicom** [comment]
Remaining criticisms as of most recent adjustments:

* Spiders will snowball out of control, completely swarming the station with sheer numbers unless the crew manages 100% eradication (This is the same problems swarmers have).
* Following from the above, spiders do not need to consume players/humans to reproduce uncontrollably due to nurses being both reproductive and not classified as enriched spawns as well as the large number of babies that are produced from each single feeding.
* Spider venom is greatly under-utilized. Only three spiders have it, and only one of them effectively uses it. The nurse and broodmother have so little venom compared to their brute damage that the venom effects will be completely negligible. 
* Vipers are stupidly powerful against most crew due to high speed and the rapidly lethal normal venom they have, and their counterbalance of being completely worthless against robotic enemies isn't fantastic imo, even though this is how they currently are.

**Comment by Fronsis** [comment]
> Remaining criticisms as of most recent adjustments:
> 
> * Spiders will snowball out of control, completely swarming the station with sheer numbers unless the crew manages 100% eradication (This is the same problems swarmers have). Multiple babies are spawned from every feeding.
> * Following from the above, spiders do not need to consume players/humans to reproduce uncontrollably due to nurses being both reproductive and not classified as enriched spawns.
> * Spider venom is greatly under-utilized. Only three spiders have it, and only one of them effectively uses it. The nurse and broodmother are not only supposed to avoid combat, but have so little venom compared to their brute damage that the venom effects will be completely negligible.
> * Vipers are stupidly powerful against most crew due to high speed and the rapidly lethal normal venom they have, and their counterbalance of being completely worthless against robotic enemies isn't fantastic imo, even though this is how they currently are.

Have you looked at /TG/ iteration of their spider gamemode? Perhaps they've found a more balanced status with other tweaks besides the core original port

To ask a question, with your proposal changes, how many laser-shots would it take to kill a Viper or most of the fragile-yet deadly spiders? I think there needs to be accountant the movement speed rework and see how easily is for a default spessmen to kite a spider and vice-versa

The snowball will also depend on how many crewmembers do spiders actually eat, since if they don't secure a player body they will not be able to hatch into stronger versions, since they would eventually get the treatment xenos get ''KILL ALL MONKEYS ON VIRO/GENE/ETC'' i'd say that would need a couple of TM to see the overall balance of spider numbers and their snowball potential once the final numbers gets settled.

**Comment by Rukofamicom** [comment]
> Have you looked at /TG/ iteration of their spider gamemode? Perhaps they've found a more balanced status with other tweaks besides the core original port

TG's balance isn't necessarily going to suit ours - the majority of our (rule-abiding) crew are usually totally unarmed, for one. There's a lot of things TG does right, but combat balance when we're so diverged from them isn't something we should use them as direct reference for imo. 

> To ask a question, with your proposal changes, how many laser-shots would it take to kill a Viper or most of the fragile-yet deadly spiders? I think there needs to be accountant the movement speed rework and see how easily is for a default spessmen to kite a spider and vice-versa

Two hits to die to a laser, three to a toolbox. On the flipside, a viper can reliably disable a single crewmate *regardless of armor* in two bites. One bite if they're lucky or the crew member is especially careless, and this is true whether they are using Spider Venom or OG venom tbh, with the biggest difference being that OG venom outright kills where spider venom paralyzes and then kills some time later. Spider venom requires assistance to finish off the envenomed crew, and also allows for the person to be dragged away with more options for treatment depending on how the teams play. 

Vipers are also the only spider fast enough to reliably dodge attacks, and especially lasers since *they are the only spider to move faster than players.*

> The snowball will also depend on how many crewmembers do spiders actually eat

This is completely incorrect. It's not crewmembers at all, or even monkeys as you implied. **anything living can be consumed to produce four new spiders**. If the consumed thing happens to be a human, one of the four babies has the option to be one of the three more powerful variants. The three normal variants may be weaker, but there is power in sheer numbers. 

I might add the event currently starts with ten spiders: Two broodmothers, six normal spawns, two enriched spawns.

**Comment by Rukofamicom** [comment]
> i'd say that would need a couple of TM to see the overall balance of spider numbers and their snowball potential once the final numbers gets settled.

100% this, nobody can know for certain without testing. I just don't want to see another midround antag that ends up being an almost guaranteed shuttle call because it's unfeasible for crew to actually eradicate the problem.

**Comment by benbot16** [comment]
The thing about vipers is that they aren't actually faster than players (still speed 0, like hunters), but their health is low enough that damage slowdown will kick in after as little as a welder hit. It's really hard testing alone how that will make up for venom against their targets, but it's going to be impractical to just run at someone, especially if they're prepared. 

I'm also not too sure about the starting spider numbers, but that's mostly there since every spider round I've has started off with a decent number (because xenobio tots), and the times I've seen admins spawn in a spider or two they usually just die immediately without help.

**Comment by Fronsis** [comment]
> This is completely incorrect. It's not crewmembers at all, or even monkeys as you implied. **anything living can be consumed to produce four new spiders**. If the consumed thing happens to be a human, one of the four babies has the option to be one of the three more powerful variants. The three normal variants may be weaker, but there is power in sheer numbers.
> 
> I might add the event currently starts with ten spiders: Two broodmothers, six normal spawns, two enriched spawns.

OH for some reason i thought it was like 1 monkey = 1 spiders, i somehow forgot how many eggs spiders actually lay irl lol, that makes sense

**Comment by itsmeow** [comment]
So, spiders are a bit OP, they spawn in too large numbers early game. There should be a curve where they get exponential growth instead of starting with 4-8 free spiders.

The damage spiders do and how much damage they take seems very balanced, but the early numbers are far too powerful.

**Comment by WrillWasTaken** [comment]
image
have you considered toning down the spider spawn amount

**Comment by benbot16** [comment]
>have you considered toning down the spider spawn amount

Assuming this was the round that just happened, no one really tried to stop the spiders, so they just kind of ate everything.

**Comment by EvilDragonfiend** [comment]
I think spiders should have a hatchling stage like how xenomorphs do

**Comment by itsmeow** [comment]
One thing I don't like is when you click to spawn an egg, you get an input list of which ones to be, right

But if someone takes it before you, you don't get that slot.

Clicking on the egg thing should guarantee you the slot unless you cancel, which then gives it back. It should also be random which egg spawn gives you the "good" one, rather than the first spawn being the good one.

Also, I would prefer a nicer UI for selecting which spider that tells you what different ones do instead of just the name. It could be made into a TGUI popup pretty easily and include stats about the spider and their abilities before picking, but that's optional.

**Comment by benbot16** [comment]
> One thing I don't like is when you click to spawn an egg, you get an input list of which ones to be, right
> 
> But if someone takes it before you, you don't get that slot.
> 
> Clicking on the egg thing should guarantee you the slot unless you cancel, which then gives it back. It should also be random which egg spawn gives you the "good" one, rather than the first spawn being the good one.
> 
> Also, I would prefer a nicer UI for selecting which spider that tells you what different ones do instead of just the name. It could be made into a TGUI popup pretty easily and include stats about the spider and their abilities before picking, but that's optional.

First part is something I should be able to do, but I'm not really familiar with TGUI-related stuff. I'll take a look at it in a bit, though.

**Comment by Rukofamicom** [comment]
It's disappointing to see the balance shifting more and more toward spiders being just brute force and overwhelming numbers instead of needing to be strategic or making good use of venom and webbing as mechanics. They look and feel like slightly weaker Xenos and have very little of the soul I was aiming for with my original concept.

**Comment by benbot16** [comment]
> It's disappointing to see the balance shifting more and more toward spiders being just brute force and overwhelming numbers instead of needing to be strategic or making good use of venom and webbing as mechanics. They look and feel like slightly weaker Xenos and have very little of the soul I was aiming for with my original concept.

I can understand that, but the original concept for the venom just really just isn't fun to play against, and I haven't had all that much success playing around with it to make it better. That's something that'll have to be tweaked more, though.

**Comment by Rukofamicom** [comment]
> I can understand that, but the original concept for the venom just really just isn't fun to play against

The solution is to make spiders less of a massive swarm, not remove the powerful venom while keeping their sheer numbers so high.

**Review by PestoVerde322 — DISMISSED** [review]
Icon seems fine, waiting for code approval.

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Dynamic 2023: Raise cost to 11

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (itsmeow) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Dynamic 2023: Weight 3

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Done

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (benbot16) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Done

**Review by itsmeow — DISMISSED** [review]
As is, I am happy enough with spiders and their function. They have good QoL and function as an antagonist, however their balance still leaves a bit to desire. I think this is better worked out with another PR than repeated testmerging as they still fundamentally work and we desperately need more midround variety. I am going to approve this, although this does not mean I think they are perfect balance wise. I would like to see more of how strategy plays out on the spider side, or what 'meta' evolves so we can work on making it more engaging and unique.

File: code/modules/events/spider_infestation.dm
Comment (itsmeow) [inline code/modules/events/spider_infestation.dm]: this should be ROLE_SPIDER

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/modules/events/spider_infestation.dm
Comment (benbot16) [reply code/modules/events/spider_infestation.dm]: Done

**Review by itsmeow — APPROVED** [review]

**Comment by itsmeow** [comment]
No one else has commented on this so I am just going to merge it, we will work on balance in the long term if needed.

**Pull request merged**
